### ISSUE TITLE
EIP-2612: 712-signed token approvals

### ISSUE THREAD

**Issue by MrChico** [ethereum/EIPs#2613]
This is a place for discussing https://github.com/ethereum/EIPs/pull/2612, which proposes an additional method `permit` for making approvals by way of signed messages rather than direct transactions

**Comment by maurelian** [comment]
IMO the 'security considerations' section should mention the front running attack that can occur when an allowance is reduced, either by `approve` or the new `permit` function. I don't think this proposal changes the risk or possible mitigations in any way, but I haven't thought hard about it. 

If you have a strong desire to do some bike shedding, you could even suggest a [specific mitigation](https://arxiv.org/pdf/1907.00903v1.pdf) for it. 😝

**Comment by Nipol** [comment]
I was looking forward to talking about this. I'd like to talk about the need for a "deadline"

What aspects of Tx Relayer do we need to take into account "deadline"? The Tx Relayer may have a response that causes tx while setting tx fee low. or no response.

And when a user signs for an interface, it requires more information. or more knowledge. Is usability not compromised? and more complexity for Tx Relayer.

So I think the deadline is unnecessary. 🤔

**Comment by maurelian** [comment]
I like `deadline`, as it solves the problem of worrying about [all the approvals](https://tac.dappstar.io/#/) one has floating out in the world. 
That said, I am sympathetic to the argument that it is inconsistent with the existing `approve` interface.

**Comment by MrChico** [comment]
Thanks @maurelian, added a comment about the approval front running attack. 
@nipol The `deadline` parameter is a way to mitigate the [free option](https://blog.althea.net/the-free-option-problem/) that the relaying party gets when receiving a signed message (it is just a deadline for the `Permit` itself, not for the approval so I don't think it solves the problem of random tokens being given allowance). In cases where the free option is not a concern, `deadline` can simply be set to `uint(-1)`, so it should be seen as an optional parameter

**Comment by MrChico** [comment]
Actually @maurelian I was just made aware of the [xDai staking token implementation of `permit`](https://etherscan.io/address/0x0Ae055097C6d159879521C384F1D2123D1f195e6#code) which I think does what exactly what you are alluding to here. It shares the same ABI as Dai and Chai, but not the same semantics. I guess we better document all three flavors in the ERC

**Comment by Nipol** [comment]
In terms of "Free Options," I think `deadline` has sufficient requirements.

But in terms of EIP-712, deadline is not convincing.
`deadline` is a difficult option for most signer to understand. at now, no example of an interface receiving a `deadline` value exists.

We can't show you the "Free Options" article every time ask for a signature. If we cannot effectively explain `deadline` to the signer and Tx Relayer, `uint256(-1)` will be used in most cases.

Furthermore, `deadline` is a consideration that does not appear in the EIP-712 specification. I think it does not have to appear as an exception in `Permit`.

If all of these signatures necessary a "deadline" option, if all of EIP-712 signatures have the same security level, I think it's better to update the EIP-712 specification. 🙂

What is the best option? What do you think?

**Comment by MrChico** [comment]
> We can't show you the "Free Options" article every time ask for a signature. If we cannot effectively explain deadline to the signer and Tx Relayer, uint256(-1) will be used in most cases.

I'm perfectly happy with `uint(-1)` being used in most cases.

>  deadline is a difficult option for most signer to understand. at now, no example of an interface receiving a deadline value exists.

There are already interfaces which deal with deadline, uniswap being one of them:
<img width="635" alt="Screen Shot 2020-04-27 at 12 21 12" src="https://user-images.githubusercontent.com/8930559/80361709-b2ca9100-8881-11ea-8b35-11e722cad059.png">

or, for `Permit` https://stablecoin.services automatically sets deadline 4 minutes into the future.

I can agree with your point that it may a difficult point for users to understand, and one that may require some EIP-712 tweaking to improve. Another ux improving solution could also be to introduce `Date` as a type to 712, to improve how it is rendered to the user.

I do think that these options fall outside the scope of this ERC though, especially since this in some sense more of a descriptive erc than a prescriptive erc; I am highlighting what has been implemented and the motivations behind it. We can certainly make tweaks in new versions, but I am mainly interested in documenting what already exists here.

**Comment by axic** [comment]
I've seen on twitter that apparently there are at least 3 different semantics contracts follow and still call it "permits". @MrChico you proposed to discuss all in the EIP/ERC.

I think an EIP should be unambiguous and promote a single way of doing things. Is it possible to do that while perhaps mentioning "semi-compatible" contracts? If not, would it make sense having 3 different EIPs?

**Comment by MrChico** [comment]
@axic I was planning on also mentioning the Stake-flavored implementation under "backwards compatibility", making that, the dai and chai implementations all belong to this " semi-compatible" class, but keep the specification as is for specificity

**Comment by axic** [comment]
Nice, sounds good!

**Comment by Nipol** [comment]
@MrChico Good! I think we've written quite a meaningful conversation in publicly.

It will be the data we or the next people will refer to when working in the future!
😉

**Comment by ptrwtts** [comment]
> The standard ERC-20 race condition for approvals applies to permit as well.

Why not take the opportunity to fix this? Otherwise, we might end up with proliferations of safePermit too. Is the assumption that the allowances will be spent immediately, so it's not actually a problem?

**Comment by MrChico** [comment]
> > The standard ERC-20 race condition for approvals applies to permit as well.
> 
> Why not take the opportunity to fix this? Otherwise, we might end up with proliferations of safePermit too. Is the assumption that the allowances will be spent immediately, so it's not actually a problem?

To be honest I don't find the approval race condition to be much of an issue in practice – it's only ever a problem when allowing addresses that can "spontaneously" spend your allowance. In most use cases, you give your allowance to a smart contract in such a way that only you can later initialize the `transferFrom` from your own address.

I'm all for people implementing `safePermit` if they feel that this is a concern, but in order to limit the scope of what I'm trying to do here I think such alternatives should be in a different ERC

**Comment by nventuro** [comment]
It wasn't clear to me from a cursory read of the EIP that the following getter is also part of the required interface, I'd make it more explicit:

```solidity
function nonces(address owner) public view returns (uint256);
```

I'd also consider changing its name to something else, since `nonces` is quite generic and non-descript.

**Comment by MrChico** [comment]
> I'd also consider changing its name to something else, since `nonces` is quite generic and non-descript.

Do you have a suggestion?

**Comment by MrChico** [comment]
Keep in mind that a change in this name would make this etc incompatible with Uniswap v2 (and even more incompatible with dai and chain)

**Comment by petejkim** [comment]
> > I'd also consider changing its name to something else, since `nonces` is quite generic and non-descript.
> 
> Do you have a suggestion?

permitNonces()

**Comment by MrChico** [comment]
Its okay for nonces to be generic, in some cases it can be an advantage. The erc doesn't specify other methods can't use `nonces` for other purpose

**Comment by MrChico** [comment]
But most of the time, this doesn't seem like a particularly important concern

**Comment by wighawag** [comment]
I think it would be useful to have a function that let smart contract know if the permit call will succeed if executed. 
For example, when a contract called via static_call need to check the validity of the signature before handing the operation to another contract/call.

This could be like :
```
function isValidPermit(address owner, address spender, uint256 value, uint256 deadline, uint8 v, bytes32 r, bytes32 s) external view returns (bool);
```

**Comment by MrChico** [comment]
> I think it would be useful to have a function that let smart contract know if the permit call will succeed if executed.
> For example, when a contract called via static_call need to check the validity of the signature before handing the operation to another contract/call.
> 
> This could be like :
> 
> ```
> function isValidPermit(address owner, address spender, uint256 value, uint256 deadline, uint8 v, bytes32 r, bytes32 s) external view returns (bool);
> ```

This is interesting. But if they were called with staticcall they wouldn't be able to execute the permit later anyway? And in the only cases I can imagine, it seems like you'd want to submit the permit whenever if possible, so wouldn't a try-catch be sufficient for those use cases? Or is there some scenario you can think of in which you would NOT want to submit a well-formed permit?

**Comment by MrChico** [comment]
Hmmm, I guess one could imagine a sort of `pure` validity checking contract which would be called with staticcall and return `true` to the original caller if all is ok, and the original caller would not be constrained to a `STATIC` environment.... 
Did you have something like this in mind? Do people actually write contracts like this?

**Comment by wighawag** [comment]
Yes, that's basically it.
GSN (https://docs.opengsn.org/learn/index.html) have this architecture where a "view" call is made first to a "paymaster" contract that can answer whether it can pay for the metatx. If it answer no, the metatx do not proceed and it is not charged. If it says yes, another call will be made where the "paymaster" can now call permit. The gas consumed here will be charged

**Comment by MrChico** [comment]
@wighawag couldn't you just as well check the validity of the permit in the "paymaster" in that case? As per the spec, you need to validate:

- The current blocktime is less than or equal to `deadline`.
- `owner` is not the zero address.
- `Token.nonces[owner]` is equal to nonce provided `nonce`.
- `r`, `s` and `v` is a valid `secp256k1` signature from `owner` of the hashed message.

all of which can be verified in a static environment.

**Comment by MrChico** [comment]
Although the `nonce` is not given as an argument in the `permit` function, it will equal `Token.nonces[owner]` if it is indeed a valid signature.

**Comment by nventuro** [comment]
I think this EIP can do without the `view` method described above. The reason why it's there in GSN is because part of that function is calling into a contract and performing an arbitrary query: by encapsulating the call there clients can get a boolean return value without having to know anything about the end recipient.

**Comment by wighawag** [comment]
@MrChico @nventuro That's exactly what I do as a **workaround**, see here : https://github.com/wighawag/gsn-playground/blob/5f40cc949031129374a48e43c8a0b71b83e52bf1/contracts/src/DAIPaymaster.sol#L101-L119

but this does not scale if my paymaster wanted to support all token that support the permit standard (unless A DOMAIN_SEPARATOR view was also part of the standard), hence why in my opinion that function should be part of the standard.

By the way in regard to EIP-712 domain. I am not sure this should be part of the standard. Contract implementer should be free to set the EIP-712 domain they see fit.

**Comment by MrChico** [comment]
> but this does not scale if my paymaster wanted to support all token that support the permit standard (unless A DOMAIN_SEPARATOR view was also part of the standard), hence why in my opinion a function like `isValidPermit` should be part of the standard.

Ah, yes you are right that this requires DOMAIN_SEPARATOR to be publicly accessible. I think requiring it to be accessible is a good idea and should be added to the spec

**Comment by nventuro** [comment]
Is that because the `version` field is unknown? All other members of the domain separator seem to be obtainble, but I'm not an expert on 712.

**Comment by MicahZoltu** [comment]
While the current mechanism for nonce management saves on gas (one storage slot per transaction rather than one storage slot per user), it means that a user cannot sign several permits and hand them out to multiple providers because inclusion order is not known at permit signing time.  It may be worth the storage costs to have the nonce be included.  An implementation can always do some tricks like storing 256 nonce-used flags in a single variable so there are 255 storage updates for every storage addition, thus greatly reducing the gas costs.

**Comment by defifuture** [comment]
I'm wondering if using `block.number` as a deadline would be a better choice than `block.timestamp`, since it's a value that can't be manipulated and you can be sure by which block the `permit()` tx must go through. What do you guys think?

**Comment by petejkim** [comment]
@MicahZoltu USDCv2 implements both permit() and [its own gas abstraction methods](https://github.com/centrehq/centre-tokens/blob/master/contracts/v2/GasAbstraction.sol) precisely for that reason. (both are great!) I'll try to work on an EIP this weekend.

**Comment by wighawag** [comment]
Is there any progress on this EIP?

I see that the DOMAIN_SEPARATOR requirement is not in the current spec : https://eips.ethereum.org/EIPS/eip-2612
@MrChico you agreed it was good to add it. Is that work in progress ?

Also, the current spec is defining the full EIP-712 message format, it should leave the domain separator parameters (
EIP712Domain) up to the implementation. Only the Permit message should be defined.
If that is already the intent, it should be clarified in the EIP.

**Comment by MrChico** [comment]
Yeah, I will address your comments shortly @wighawag

**Comment by k06a** [comment]
@defifuture: https://consensys.github.io/smart-contract-best-practices/recommendations/#the-15-second-rule

> The 15-second Rule¶
> The Yellow Paper (Ethereum's reference specification) does not specify a constraint on how much blocks can drift in time, but it does specify that each timestamp should be bigger than the timestamp of its parent. Popular Ethereum protocol implementations Geth and Parity both reject blocks with timestamp more than 15 seconds in future. Therefore, a good rule of thumb in evaluating timestamp usage is:
> 
> If the scale of your time-dependent event can vary by 15 seconds and maintain integrity, it is safe to use a block.timestamp.

**Comment by Eenae** [comment]
I suggest clarifying what `bytes(x)` means in the solidity pseudocode snippet. Because implementations tell us that `address`es are getting padded up to 32 bytes (by `abi.encode`) - that is not obvious from the spec.

Btw, in an earlier version of the spec I saw `abi.encodePacked` which yields different results. I understand that addresses padding came from EIP 712, but still, clarification will be beneficial IMO.

**Comment by juniset** [comment]
I'm a bit late to the party but it would be nice if this EIP was compatible with EIP-1271 (https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1271.md) and also accepts signatures from smart-contract wallets. 

Smart-contract wallets (Argent, Gnosis, Authereum, etc) natively support gas abstraction but interactions with dapps that rely on `permit` fail through e.g. WalletConnect because the signature is not being validated. We either need to ask dapps to support 2 different flows (which is bad) or `permit` should accept signatures from EOAs and smart-contract wallets.

**Comment by Ro5s** [comment]
This may be naive, but can this implementation work with NFT/erc721 where ‘value’ might represent a ‘tokenId’? (Or in simple adaptation, this param would be explicit?)

**Comment by remote-gildor** [comment]
I think the `permit()` method should return a `bool` to go in line with other ERC-20 methods:

```
function permit(address owner, address spender, uint value, uint deadline, uint8 v, bytes32 r, bytes32 s) external returns (bool success)
```

What do you think?

**Comment by MrChico** [comment]
@remote-gildor is there any particular reason you need this `bool` return? Imho the `bool` return in ERC20 was largely a mistake and I don't see a reason for including it here.

**Comment by remote-gildor** [comment]
Hey @MrChico 🙂 I was looking at this list of unexpected ERC-20 behaviours and saw the missing return value section: https://github.com/xwvvvvwx/weird-erc20#missing-return-values 

Then I googled a bit more and found this article (Missing Return Value bug): https://medium.com/coinmonks/missing-return-value-bug-at-least-130-tokens-affected-d67bf08521ca 

Not sure how exactly could this be exploited with the `permit()` method, but I guess it doesn't hurt if the method returns a bool.

**Comment by albertocuestacanada** [comment]
I've just written some integration with USDC v2 eip2612 implementation. USDC v2 uses the "2" version in the domain separator, but there isn't any way of getting the version out of the contract, thus making automated integration impossible.

I'd suggest that since the eip712 domain separator of eip2162 requires a `version`, a getter for it should be part of the specification.

[Potential implementation](https://github.com/yieldprotocol/yield-utils/blob/6457ec5d1cd4b4ffaa9dd20f1efa6cf7055bb8c9/contracts/token/ERC20Permit.sol#L40), that allows overriding of the version by inheriting contracts.

**Comment by MrChico** [comment]
Hmmm, as it stands, this specification does not specify the form of the domain separator. Anything that satisfies https://eips.ethereum.org/EIPS/eip-712#definition-of-domainseparator will do. This EIP mandates the `DOMAIN_SEPARATOR()` to be public though. It would be hard to go further without mandating a format on the DOMAIN_SEPARATOR.
I suppose you are suggesting that the `DOMAIN_SEPARATOR()` should be standardized here as well, and that all of its fields should be available in publicly callable functions?

**Comment by albertocuestacanada** [comment]
I didn't realize that I'm supposed to pull the `DOMAIN_SEPARATOR` of the contract, instead of computing it off-chain. That makes things easier.

**Comment by petejkim** [comment]
`version` is still required if you want to use `eth_signTypedData[vX]`, so I think it's a good idea to add a `version()` function.

**Comment by alfredolopez80** [comment]
> I'm a bit late to the party but it would be nice if this EIP was compatible with EIP-1271 (https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1271.md) and also accepts signatures from smart-contract wallets.
> 
> Smart-contract wallets (Argent, Gnosis, Authereum, etc) natively support gas abstraction but interactions with dapps that rely on `permit` fail through e.g. WalletConnect because the signature is not being validated. We either need to ask dapps to support 2 different flows (which is bad) or `permit` should accept signatures from EOAs and smart-contract wallets.

it was resolved, with Open Zeppelin Library available in version 4.1
https://docs.openzeppelin.com/contracts/4.x/api/utils#SignatureChecker

@juniset

**Comment by frangio** [comment]
> I didn't realize that I'm supposed to pull the `DOMAIN_SEPARATOR` of the contract, instead of computing it off-chain. That makes things easier.

@alcueca I don't think this is right. If you just pull the domain separator from the contract you have no way of verifying that it is the appropriate chain id. Some other solution is needed, but it sounds like a broader problem with EIP712 rather than in ERC20-permit.

**Comment by juniset** [comment]
> > I'm a bit late to the party but it would be nice if this EIP was compatible with EIP-1271 (https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1271.md) and also accepts signatures from smart-contract wallets.
> > Smart-contract wallets (Argent, Gnosis, Authereum, etc) natively support gas abstraction but interactions with dapps that rely on `permit` fail through e.g. WalletConnect because the signature is not being validated. We either need to ask dapps to support 2 different flows (which is bad) or `permit` should accept signatures from EOAs and smart-contract wallets.
> 
> it was resolved, with Open Zeppelin Library available in version 4.1
> https://docs.openzeppelin.com/contracts/4.x/api/utils#SignatureChecker
> 
> @juniset

@alfredolopez80 Not really since the current spec still mentions `r, s and v is a valid secp256k1 signature from owner of the message:` which suggests that the owner is an EOA.

We are pushing for OZ to use `https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/cryptography/SignatureChecker.sol` in their implementation of `permit` (see https://github.com/OpenZeppelin/openzeppelin-contracts/issues/2845) but they want to be sure that they are compliant with the spec.

It would be great if the spec was making it clear that `r,s,v` is a valid secp256k1 signature or a valid signature according to EIP1271.

**Comment by k06a** [comment]
Really sad that EIP have `r`, `s`, `v` arguments instead of `bytes signature`. Having string would allow to use [`SignatureChecker.sol`](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/cryptography/SignatureChecker.sol) abstraction on top of EOA/SC signatures and to use [EIP-2098 compact signatures](https://eips.ethereum.org/EIPS/eip-2098)

**Comment by frangio** [comment]
Yes, `bytes signature` would be better, but unlikely to change at this point. But even with `r,s,v` it would be possible to use `SignatureChecker` for ECDSA-style EIP1271 signatures, if the EIP allows it, which is a backwards compatible improvement.

**Comment by axic** [comment]
@MrChico this EIP is now in stagnant, but many projects seem to depend on it. I think it would be important getting this to `Final` to avoid even more fragmentation. Different behaviour can then be codified in a new EIP if a different interface is sought.

**Comment by MrChico** [comment]
@axic what is missing to Finalize it in your opinion?

**Comment by axic** [comment]
I haven't 100% checked the content, but from a process perspective it is a few steps. Now that it lapsed into Stagnant, it needs to be brought back to `Draft`/`Review` (it is agreed by the editors that it can go directly to `Review`). One you rallied enough relevant parties (i.e. the people who engaged here), it can be moved to `Last Call` for (at least 2 weeks) and if no substantial disagreement/comments come up it becomes `Final`.

i.e. your next step is to create a PR moving it to `Draft` or `Review`, but you need to make sure the format matches the requirements of EIP-1, which has undergone several changes since you created the draft (new preamble fields, different section requirements, etc.)

**Comment by MrChico** [comment]
I tried moving this ERC to last call before https://github.com/ethereum/EIPs/pull/3308, which was met with a series of bike shedding and bureaucratic problems relating to the EIP dependencies. I agree that it would be valuable to mark this EIP as `Final`. As far as I am concerned the spec is Final. I don't feel particularly motivated to address the numerous comments raised in the linked PR, which really have more to do with formatting rather than the actual spec

**Comment by axic** [comment]
Oh I see. That is very unfortunate.

**Comment by nventuro** [comment]
@axic could someone else address the comments to have this be finalized? The entire EIP space is ridden with 'draft' specs that people are afraid to build on, it'd be useful to mark at least some of the most widely used (like this one) as 'final'.

@MicahZoltu's [comments](https://github.com/ethereum/EIPs/pull/3308#pullrequestreview-602729299) seem simple enough to address, except for the request to move 712 to a 'final' stage. This is another of these huge forever-draft EIPs that still see widespread use but remain unfinalized despite [pretty much no discussion happening](https://ethereum-magicians.org/t/eip-712-eth-signtypeddata-as-a-standard-for-machine-verifiable-and-human-readable-typed-data-signing/397/28). Is not linking to 712 really a hard requirement?

**Comment by frangio** [comment]
Seems like we should start by finalizing EIP-712 and then move on to this one.

**Comment by MicahZoltu** [comment]
> Is not linking to 712 really a hard requirement?

Yes.  As I mentioned in the comments, one option is to just create a copy of that EIP (with a new number) and actually push it through the process.  It would be *better* to get the authors to push 712 through the process or get the authors of 712 to hand over authorship to you so you can push it through the process, but creating a clone EIP is always a backup option (caveat being that you'll get a new number).

**Comment by wuya666** [comment]
It seems the current design of the `nonces` makes the permit functionality strictly sequential for a specific `owner`, ie. an owner cannot sign multiple approval permits to multiple spenders at the same time, as one spender calling the permit function will invalidate all the other signatures. I'm not too sure if there will be some severe security issues if I change the `nonces[owner]` mapping into `nonces[owner][spender]` mapping so that one owner can sign multiple permits to different spenders at a time?

**Comment by k06a** [comment]
@wuya666 it's possible to keep both:

```solidity
require(nonce >= nonces[owner], "nonce is too old or cancelled);
require(nonce == nonces[owner][spender]++, "nonce is too old");

function cancelNoncesLessThan(uint256 nonce_) external {
    require(nonce_ > nonces[msg.sender] + 1_000_000;
    nonces[msg.sender] = nonce_;
}
```

**Comment by k06a** [comment]
But maybe we need to manage nonces on higher level, in EIP-712 ...

**Comment by favealo** [comment]
🙃

**Comment by k06a** [comment]
Or better have layer between EIP712 and EIP2612 to describe cancellable signature-based operations.

**Comment by frangio** [comment]
I believe permits are usually signed and immediately used, and in this case sequential nonces are not a problem.

**Comment by wuya666** [comment]
> I believe permits are usually signed and immediately used, and in this case sequential nonces are not a problem.

well, it depends on how fast the transaction is mined, which can take seconds to hours (to forever) depending on the gas price provided and the network congestion situation. For standard ERC20 approvals, it's not a problem since I can just queue multiple approval transactions with incrementing transaction nonces and they'll be mined in that order when the gas price is accepted by miners, but for permitted approvals, there's no control over which one is mined first, so basically I need to be sure the last approval is mined, or the deadline is passed, before signing the next approval.

**Comment by wuya666** [comment]
> @wuya666 it's possible to keep both:
> 
> ```solidity
> require(nonce >= nonces[owner], "nonce is too old or cancelled);
> require(nonce == noncesFromTo[owner][spender]++, "nonce is too old or cancelled");
> ```
> 
> ```solidity
> function cancelNoncesLessThan(uint256 nonce_) external {
>     require(nonce_ > nonces[msg.sender] + 1_000_000);
>     nonces[msg.sender] = nonce_;
> }
> 
> function cancelNoncesFromTo(address spender, uint256 nonce_) external {
>     require(nonce_ > noncesFromTo[msg.sender][spender] + 1_000);
>     noncesFromTo[msg.sender][spender] = nonce_;
> }
> ```

@k06a I don't think it will work by checking the ">=" condition? Let's say I have signed 100 approvals for 100 people out there, with nonces being 0 to 99, how should I increment the nonce[owner] counter after one of them called the permit function?

**Comment by k06a** [comment]
@wuya666 ‘nonces’ are used for cancelling only and not requires incrementing at all.

**Comment by wschwab** [comment]
> Seems like we should start by finalizing EIP-712 and then move on to this one.

I'm actively working on trying to get this done, fwiw

One more thing: If it's too late in the game to ask, I totally understand, but wouldn't adding EIP-165 be a good idea? That way platforms would have an easy check if an ERC20 contract has `permit` or not. I'd be happy to PR myself to add this if @MrChico thinks it's a good idea.

**Comment by frangio** [comment]
Does anyone have stats on how many tokens already support permit? EIP-165 does sounds like it makes sense for this ERC, if it hadn't been in production for such a long time already.

**Comment by wuya666** [comment]
> @wuya666 `nonces` In my case are used for cancelling only and not requires incrementing at all.

@k06a without incrementing `nonce` in the contract, the signature can be used to do permitted approvals multiple times as long as the deadline is not reached right? Isn't that a big security hole?

**Comment by k06a** [comment]
@wuya666 I meat I use same non in multiple ways. One of them increments, another ways is for bulk cancelling.

**Comment by vrypan** [comment]
Regarding the deadline. We are starting to see cases where an old `approve()` is used to steal tokens from users: https://twitter.com/dingalingts/status/1470095710888808449 

I think it should be a deadline for `transferFrom()` and not for `permit()`.  

I have been drafting an EIP to address this, but do you think that we could/should integrate these ideas in 2613?
Here is the draft: https://github.com/vrypan/EIPs/blob/master/EIPS/eip-draft_time_limited_token_allowances.md

**Comment by williamkmlau** [comment]
> It seems the current design of the `nonces` makes the permit functionality strictly sequential for a specific `owner`, ie. an owner cannot sign multiple approval permits to multiple spenders at the same time, as one spender calling the permit function will invalidate all the other signatures. I'm not too sure if there will be some severe security issues if I change the `nonces[owner]` mapping into `nonces[owner][spender]` mapping so that one owner can sign multiple permits to different spenders at a time?

Hi all, I actually just wanted to throw some support for the nonce changes suggested by @wuya666 above during the discussions in November. 

I've worked on several DeFi projects in the past 18 months or so and the existing implementation had always been sufficient, as most use cases only involve a user signing permits to be processed by platform contracts, which is unlikely to be submitted in parallel or in bulk.

However since Q3/Q4 2021 with the NFT boom, I've been working on multiple GameFi projects and have seen major needs instead for platform contracts to sign permits so individual users can collect/claim tokens based on off-chain data stored on game servers. 
This method is used as an alternative to having to oracle data on-chain, or directly initiating token transfers from a server managed wallet. This is (mostly) because most game projects would want to off-load the gas costs associated with the token transfers to users (without having to worry about users abusing reward collection mechanisms).

The current implementation of `nonces[owner] ` is insufficient to handle signing permits to multiple users simultaenously.
A change to `nonces[owner][spender]` solves this problem and is the method I've been using in my recent GameFi projects.

**Comment by frangio** [comment]
The [recent Multichain/AnySwap vulnerability](https://media.dedaub.com/phantom-functions-and-the-billion-dollar-no-op-c56f062ae49f) that put almost half a billion at risk was due to a combination of permit, infinite approvals, and "phantom functions" (i.e. a fallback function that accepts any function call).

A literal reading of the EIP shows that the last part is not strictly necessary. A compliant implementation of `permit` isn't required to revert if the permit is invalid or old, and this could lead to the same vulnerability that was found there.

I want to suggest adding to the specification that `permit` must revert if the signature is invalid.

**Comment by duckki** [comment]
I have the same concern as @frangio. Since `permit` returns no value, the caller contract can't detect if the `permit` was successful or the ERC-20 contract simply didn't implement `permit`.
One way to mitigate this is to change `permit` to return a bool value indicating the success/failure, instead of relying on a token contract to implement `permit` and revert. In this way, at least, the caller contract can detect token contracts that didn't implement `permit`. Also, a wrapper function (say, `safePermit()`) could be implemented analogous to safeTransfer/safeTransferFrom.

**Comment by frangio** [comment]
Adding a boolean return value would be a breaking change for the many implementations already in production, so I don't think it's a good idea. But reverting on invalid signatures seems to be what current implementations are doing so we should just make that an explicit part of the specification.

A wrapper function could check that the account nonce is incremented, as suggested in https://github.com/OpenZeppelin/openzeppelin-contracts/issues/3145.

**Comment by duckki** [comment]
Indeed, OpenZeppelin/openzeppelin-contracts#3145 is way to go.

**Comment by k06a** [comment]
@frangio increasing amount of returning data is not actually “breaking” change. Solidity reverts when returned data size is less than expected by interface.

**Comment by k06a** [comment]
@frangio increasing amount of returning data is not actually “breaking” change. Solidity reverts when returned data size is less than expected by interface.

**Comment by frangio** [comment]
Depending on the specific change we're talking about, that may be true in in one direction (future implementations compatible with current users) but not in the other (future users not compatible with current implementations).

I mean that all current users of permit would work fine with a hypothetical permit implementation that returns a value. But if the EIP is changed so that a return value is required, future users of permit that expect a return value would break with all the current implementations that don't return anything.

The way around this is to specify the return value as optional. But this is not natively supported by Solidity at least, so it would be more complicated to use.

**Comment by k06a** [comment]
@frangio agree. Good thing that we already have experience in handling this: SafeERC20

**Comment by gonbad** [comment]
It would be better to put `nonce` in the `permit` function arguments. This way we can use a random number as `nonce` and prevent concurrency issues.

**Comment by gonbad** [comment]
One use-case of these contracts could be cheques; so it would be nice to have a `form_timestamp` argument.

**Comment by gonbad** [comment]
It's sad that we don't have a `PKCS` standard for encoding `r`, `s` and `v` to a single `bytes` argument.

**Comment by TimDaub** [comment]
The current document links to files of the Uniswap GitHub organization:

- https://eips.ethereum.org/EIPS/eip-2612#test-cases: https://github.com/Uniswap/uniswap-v2-core/blob/master/test/UniswapV2ERC20.spec.ts.
- https://eips.ethereum.org/EIPS/eip-2612#reference-implementation: https://github.com/Uniswap/v2-core/blob/master/contracts/UniswapV2ERC20.sol

A few notes:

- When publishing EIP-4973 for DRAFT, I was asked to put such files as CC0 licensed in the EIPs repository's assets folder.
- Uniswap/v2-core is GPL3
- The links aren't permanent but they point to Uni's "master" directory, meaning those files could change over time
- Additionally, I think it's in the interest of the EIP process to be self-referential. So links pointing to external sources should be considered invalid. Otherwise, I'm happy to also host all of EIP-4973's files on my STARTUPS GitHub so that everyone can see what awesome work I do etc.

Few more notes:

- `Discussions-To` points to this GitHub issue but should point to Ethereum Magicians
- There are other links referencing external sources. I don't know what is the general line of thinking in the EIP process, but IMO those links should at least be automatically snapshotted with e.g. the webarchive or idk uploaded to IPFS for posterity.

**Comment by MicahZoltu** [comment]
I agree that the external links should be removed.

This EIP was crafted before we moved discussions to Ethereum Magicians though, and we are generally grandfathering existing discussion links in GitHub to avoid discontinuity in discussions.  If the authors want to move the discussion to Ethereum Magicians that is great, but not required.

**Comment by phil-ociraptor** [comment]
> TL;DR: EIP appears stuck mostly because: **1)** EIP-712 itself not yet finalized, **2)** official support for EIP-1271 not agreed upon, **3)** debate on if `nonces` should be `nonces[owner]` vs `nonces[spender][owner]`. 
>
> Suggestion is to finalize EIP-712 as is, as well as ERC-2612 as is, but also introduce an optional alternative i.e. `ERC-2612A`, that is more feature rich (i.e. has EIP-1271 support and `nonces[spender][owner]`). Unblocks us and can move forward on this very popular EIP

This EIP has been outstanding for over 2 years, despite having many implementations in the wild already. The longer the standard is not finalized, the longer there will be ambiguity, and create more reluctance to adopt this standard, which most people seem to want. 

Many new tokens are already implementing EIP-2612, but the draft status of this EIP is a deterrent for adoption.

We have already resolved security concerns about tokens that do not revert, by recommending calling contracts use `safePermit` (https://github.com/OpenZeppelin/openzeppelin-contracts/issues/3145) and specifying that tokens _should_ revert if they wish to correctly implement `EIP-2612` (https://github.com/ethereum/EIPs/pull/4728/files) (but contracts should still use `safePermit`)

 It seems to me that the following points are blockers:
- EIP-712 is in Last Call, but not Final (as a representative from 0x - where EIP-712 originated, I can _try_ to herd the cats on this one)
- `nonces[owner]` vs `nonces[spender][owner]`: I agree that the `nonces[spender][owner]` is preferable, however, this will result in a breaking change to the interface that exists for many tokens already deployed
- whether `EIP-1271` support should be required such that smart contract wallets can benefit from `permit` support

It seems to me that we may need another ERC, i.e. `ERC-2612A` (or something better named, suggestions welcome), which codifies the `nonces[spender][owner]` and EIP-1271 behavior, but that should be a _variation_ of the EIP-2612 standard. This variation might need some additional functions/flags (i.e. a function that identifies this contract as a 2612A variant, such that `safePermit` knows to type switch on the correct variation when checking the `nonce` increase)

This allows us to:
- get ERC-2612 finalized, such that all token contracts that already implement this can get "retroactive" status as being fully compliant with a _finalized_ EIP, instead of being in the awkward position of implementing a standard that is still in review
- (stretch goal) produce a second ERC, i.e `ERC-2612A` for the `nonces[spender][owner]` and EIP-1271 behavior

The names of `ERC-2612` and `ERC-2612A` can be improved to indicate that one is the legacy, widespread version and the other is the new and preferred version, but this allows us to push ahead. 

How does this sound to everyone here, and am I missing anything?

As an aside, EIP-1271 support needs to be carefully considered, and I'm ok with dropping support for it in the alternative impl. Calling an arbitrary contract in the context of granting approvals seems like it's tempting fate - but with careful consideration, maybe we think carefully of the attack vectors and its ok to add

Once this is done, let's get this finalized for NFTs too 👀  https://eips.ethereum.org/EIPS/eip-4494 cc: @wschwab

**Comment by frangio** [comment]
I've already been doing the work to move EIP-712 to Final, by moving it to Last Call just a few weeks ago and going through the requisite editorial process. The Last Call period ends on Monday and the EIP should move to Final then.

Once that happens, I believe EIP-2612 should be moved to Last Call.

I also recently proposed [EIP-5267](https://eips.ethereum.org/EIPS/eip-5267) which tackles another blocker for EIP-712 adoption. Those interested should review that and give feedback and I will try to finalize it soon as well.

As for the nonces issue, I agree that at this point EIP-2612 is de facto Final and shouldn't be changed in any backwards incompatible way. I would support a "permit v2" ERC that could be implemented alongside EIP-2612 or on its own. I'd add support for EIP-1271 signatures to the wishlist.

**Comment by phil-ociraptor** [comment]
Amazing - EIP-5267 will solve one of the biggest pain points around using EIP-712: "how in the world do I generate the domain separator?".

Thank you for all the great work! 

As an aside, I'd love to spec out a new reference impl of WETH9 for new EVM chains and L2s such that they can support `permit` - I think the major decision point there will be if this contract goes w/ `permit` support of this "permit v2"

**Comment by skaille** [comment]
https://www.dropbox.com/s/f9j8up93j7mtjzr/RNFetchBlobTmp_bypu526ycsnu4tkliumi6?dl=0

**Comment by frangio** [comment]
Proposing to move EIP-2612 to Last Call with deadline September 6th 2022. https://github.com/ethereum/EIPs/pull/5506/files

**Comment by frangio** [comment]
**This EIP is now in Last Call until October 8th.**

**Comment by k06a** [comment]
What about EIP-2098 (Compact Signatures Representation)?

**Comment by frangio** [comment]
@k06a What is your proposal?

**Comment by k06a** [comment]
Let’s store `v` component of the signature in top bit of `s` component since it’s always empty. This would allow to save 32 bytes of calldata (including padding)

**Comment by frangio** [comment]
In my opinion we should consider EIP-2612 frozen and start working on an improved EIP that extends this one, as suggested by @phil-ociraptor above.

Note that using EIP-2098 is at odds with EIP-1271 which is another thing that has been proposed as a needed improvement to permits.

**Comment by TimDaub** [comment]
why? In Eup-4973 we use both 1271 any 2098 and I didn't notice anything at odds

**Comment by frangio** [comment]
In EIP-1271 the signature has no specified semantics, it should be seen as a blob, so it's not correct to apply the EIP-2098 encoding to it since it's specific to ECDSA. Unless perhaps it's only applied as an optimization for the EOA case? I need to think about that more.

**Comment by TimDaub** [comment]
Cross linking to EIP-4973 on Magicians here for reference: https://ethereum-magicians.org/t/eip-4973-account-bound-tokens/8825/146?u=timdaub

**Comment by Pandapip1** [comment]
EIP-2612 need not be considered frozen. You can always change the spec until it's final (as was done with EIP-712).

**Comment by frangio** [comment]
That is true in theory, but in practice making normative changes to an EIP that's been widely deployed can be very problematic so I think it should be avoided.

**Comment by Pandapip1** [comment]
> That is true in theory, but in practice making normative changes to an EIP that's been widely deployed can be very problematic so I think it should be avoided.

Fair enough. It should be pushed to final regardless, though.

**Comment by TimDaub** [comment]
I think we should go full no-mercy here. If ppl implemented an EIP in non-final status that's their fault and I don't think we should give into "oh but they have implemented it already." We're EIP editors and our legitimacy comes from executing EIP-1 to the best of our capabilities.

I know this will make some enemies but I believe it sends the right signal to implementers to make breaking changes to EIP-2612.

Or alternatively move EIP-2612 into withdrawn unless someone steps up and moves it into final. And then create an updated EIP with the breaking changes.

**Comment by frangio** [comment]
> unless someone steps up and moves it into final. And then create a new EIP with the breaking changes.

This is exactly what was proposed and what is happening. I opened the PR to move to Final yesterday: https://github.com/ethereum/EIPs/pull/5782

**Comment by rube-de** [comment]
any reason (beside the CI error of missing ethereum-magicians.org thread) this is not Final now?

**Comment by Pandapip1** [comment]
This has reached Final.

**Comment by invocamanman** [comment]
Hi! i might be late, but i want to propose to add:
`function DOMAIN_TYPEHASH() external view returns (bytes32)`
Since

**Comment by invocamanman** [comment]
Hi! i might be late, but i want to propose to add:
`function DOMAIN_TYPEHASH() external view returns (bytes32)`
The rationale behind is that there are right know 2 typical ways of creating the DOMAIN_TYPEHASH:
- The one proposed on EIP-2612 as we can see in [WETH10 impl](https://github.com/WETH10/WETH10/blob/main/contracts/WETH10.sol#L55)
  - ` keccak256("Permit(address owner,address spender,uint256 value,uint256 nonce,uint256 deadline)")`
- And another one that basically retrieve the version field, which is used for example in  [UNI mpl](https://github.com/Uniswap/governance/blob/master/contracts/Uni.sol#L53)
  -  `keccak256("EIP712Domain(string name,uint256 chainId,address verifyingContract)")`
 
With the current standard view functions,  a front-end cannot know which type of `Domain` uses the contract that want to interact with `permit`, and therefore cannot create properly the signature.
With this standard view function which some ERC20 permit contracts implements, a FE is able to discern between these 2 implementations. 
Currently some of contracts implement this as `DOMAIN_TYPEHASH` another ones `EIP712DOMAIN_HASH`  and another ones do not implemented at all.
I think it will help to add it to the standard to avoid confusions.

**Comment by frangio** [comment]
@invocamanman The domain hashes are not enough. Please see [EIP-5267](https://eips.ethereum.org/EIPS/eip-5267) for a proposal to tackle this problem.

**Comment by Pandapip1** [comment]
@lightclient if you meant to unlock, fixed.
